client/src/shutdown/components/ShutdownButton.tsx

```diff
@@ -0,0 +1,67 @@
+import * as React from "react";
+import { connect } from "react-redux";
+import { Button, Header, Icon, Modal } from "semantic-ui-react";
+import { handleSubmit } from "../api";
+import { RootReducer } from "../../store";
+
+const mapStateToProps = (state: RootReducer) => {
+    return {
+        channel: state.channelStatus.status,
+    };
+};
+
+type MergedProps = ReturnType<typeof mapStateToProps>;
+
+class ShutdownButton extends React.Component<MergedProps> {
+    public state = {
+        modal: false,
+        shutdown: false,
+    };
+
+    public modalOpen = () => {
+        this.setState({ modal: true });
+    };
+
+    public modalClose = () => {
+        this.setState({ modal: false });
+    };
+
+    public handleShutdown = () => {
+        this.setState({ shutdown: true });
+        handleSubmit();
+    };
+
+    public handleModal() {
+        const { shutdown } = this.state;
+        const { channel } = this.props;
+
+        if (channel === "disconnected" && shutdown) {
+            this.setState({ shutdown: false });
+            this.modalClose();
+        }
+    }
+
+    public render() {
+
+        this.handleModal();
```

Comment:
is it okay to use function inside `render` ?

Reply:
Good call - I think calling `setState` in `render` can be a problem, but doesn't have to be (one needs to be careful not to create infinite loops, as calling `setState` will result in a re-render).